From 55906e135562185edb40845f9c97966c780ff5c2 Mon Sep 17 00:00:00 2001
From: gbhakteX <gurudattax.bhakte@intel.com>
Date: Thu, 26 Jul 2012 18:01:27 +0530
Subject: Mediaserver crash while trying to play unsupported container via RTSP.

BZ: 20348

The RTSP server didn't send the rtpmap attribute which is required during setupTrack(),
we now check if the number of rtpmap attributes and media-formats obtained as response
to DESCRIBE request are equal. If unequal, it results in parsing error.

Category: aosp improvement
Domain: AUDIO_MEDIA-Audio Framework
Origin: Internal
Upstream-Candidate: Yes

Change-Id: I949843190dae5e6f4eefa2e7887ea13afb734800
Orig-Change-Id: I549c8d05b8b917e57622308c9fd8eddde326a10d
Signed-off-by: gbhakteX <gurudattax.bhakte@intel.com>
Reviewed-on: http://android.intel.com:8080/77900
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Reviewed-by: Kandasamy, Muthukumar <muthukumar.kandasamy@intel.com>
Reviewed-by: M, Arulselvan <arulselvan.m@intel.com>
Tested-by: M, Arulselvan <arulselvan.m@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
Signed-off-by: Arulselvan M <arulselvan.m@intel.com>
---
 media/libstagefright/rtsp/ASessionDescription.cpp |   12 ++++++++++++
 1 files changed, 12 insertions(+), 0 deletions(-)

diff --git a/media/libstagefright/rtsp/ASessionDescription.cpp b/media/libstagefright/rtsp/ASessionDescription.cpp
index a9b3330..2c71437 100644
--- a/media/libstagefright/rtsp/ASessionDescription.cpp
+++ b/media/libstagefright/rtsp/ASessionDescription.cpp
@@ -53,6 +53,7 @@ bool ASessionDescription::parse(const void *data, size_t size) {
     mFormats.push(AString("[root]"));
 
     AString desc((const char *)data, size);
+    unsigned int rtpmapAttributeCount = 0;
 
     size_t i = 0;
     for (;;) {
@@ -102,6 +103,11 @@ bool ASessionDescription::parse(const void *data, size_t size) {
                 } else {
                     key.setTo(line, 0, colonPos);
 
+                // Ensure rtsp server has sent rtpmap attribute for media format
+                if (key == "a=rtpmap") {
+                    rtpmapAttributeCount++;
+                }
+
                     if (key == "a=fmtp" || key == "a=rtpmap"
                             || key == "a=framesize") {
                         ssize_t spacePos = line.find(" ", colonPos + 1);
@@ -158,6 +164,12 @@ bool ASessionDescription::parse(const void *data, size_t size) {
         i = eolPos + 1;
     }
 
+    // the string "[root]" is always present in mFormats. After this, ensure that
+    // one rtpmap attribute is received for each media format, as response to DESCRIBE request
+    if (rtpmapAttributeCount != (mFormats.size() - 1)) {
+        return false;
+    }
+
     return true;
 }
 
-- 
1.7.4.1

